README.md
# oscarwilder <img src="https://pachamaltese.github.io/oscarwilder/hexicon.svg" width=150 align="right" alt="hexicon"/>

<!-- badges: start -->
[![R build
status](https://github.com/pachamaltese/oscarwilder/workflows/R-CMD-check/badge.svg)](https://github.com/pachamaltese/oscarwilder/actions?workflow=R-CMD-check)
[![Codecov test
coverage](https://codecov.io/gh/pachamaltese/oscarwilder/branch/master/graph/badge.svg)](https://codecov.io/gh/pachamaltese/oscarwilder?branch=master)
[![Lifecycle: maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
<!-- badges: end -->

## An R Package for Oscar Wilde's Complete and Presented Plays

> *"Those who find beautiful meanings in beautiful things are the cultivated. For these there is hope. They are the elect to whom beautiful things mean only beauty."* Oscar Wilde.

This package is a fork of [janeaustenr](https://github.com/juliasilge/janeaustenr) and provides access to the full texts of Oscar Wilde's 7 completed, presented plays. The UTF-8 plain text for each novel was sourced from [Project Gutenberg](https://www.gutenberg.org/), processed a bit, and is ready for text analysis. Each text is in a character vector with elements of about 70 characters. The package contains:

* `vera_or_the_nihilists`: *Vera; Or, The Nihilists*, published in 1880
* `salome`: *Salome*, published in 1881
* `the_dutchess_of_padua`: *The Dutchess of Padua*, published in 1881
* `lady_windermeres_fan`: *Lady Windermere's Fan*, published in 1882
* `a_woman_of_no_importance`: *A Woman of No Importance*, published in 1883
* `an_ideal_husband`: *An Ideal Husband*, published in 1895
* `the_importance_of_being_earnest`: *The Importance of Being Earnest*, published in 1895

There is also a function `wilde_plays()` that returns a tidy data frame of all 7 plays. 

Users should be aware that there are some differences in usage between the novels as made available by Project Gutenberg. For example, "anything" vs. "any thing", "Mr" vs. "Mr.", and using underscores vs. all caps to indicate italics/emphasis. 

The texts were obtained by using `gutenbergr::` and some additional text cleaning was made by using `stringr::`.

## Installation

You can install the released version of lp from Github with:

``` r
source("https://install-github.me/pachamaltese/oscarwilder")
```

## How to Use This Package

For some ideas on getting started with analyzing these texts, see Julia Silge's excellent post
[blog post on sentiment analysis of Austen's novels](https://juliasilge.com/blog/if-i-loved-nlp-less/). For help within R, try `?salome` or similar for getting started with the data sets.

This project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS.md
# oscarwilder 0.1.0

* Initial release of Oscar Wilde's 7 completed, presented plays


tests/testthat.R
library(testthat)
library(oscarwilder)

test_check("oscarwilder")


tests/testthat/test_wilde_plays.R
# tests for austen_books function

context("Tidy dataframe for books")

suppressPackageStartupMessages(library(dplyr))

test_that("factor order is correct", {
        d <- wilde_plays()
        expect_equal(levels(d$play)[1], "Vera; Or, The Nihilists")
        expect_equal(levels(d$play)[7], "The Importance of Being Earnest")
})

test_that("tidy frame for Wilde plays is right", {
        d <- wilde_plays() %>% 
                group_by(play) %>%
                summarise(total_lines = n())
        expect_equal(nrow(d), 7)
        expect_equal(ncol(d), 2)
        # the factor levels still in the right order?
        expect_equal(as.character(d$play[1]), "Vera; Or, The Nihilists")
        expect_equal(as.character(d$play[7]), "The Importance of Being Earnest")
        # The Importance of Being Earnest has fewer lines than An Ideal Husband?
        expect_lt(d$total_lines[7], d$total_lines[6])
})


R/novels.R
#' Jane Austen's Complete Novels
#' 
#' This package contains the complete text of Jane Austen's 6 completed, 
#' published novels, formatted to be convenient for text analysis.
#' @docType package
#' @name janeaustenr
#' @aliases janeaustenr janeaustenr-package
NULL

#' The text of Jane Austen's novel "Sense and Sensibility"
#'
#' A dataset containing the text of Jane Austen's 1811 novel "Sense and 
#' Sensibility". The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/161}
#' @format A character vector with 12262 elements
"sensesensibility"

#' The text of Jane Austen's novel "Pride and Prejudice"
#'
#' A dataset containing the text of Jane Austen's 1813 novel "Pride and 
#' Prejudice". The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/1342}
#' @format A character vector with 12447 elements
"prideprejudice"

#' The text of Jane Austen's novel "Mansfield Park"
#'
#' A dataset containing the text of Jane Austen's 1814 novel "Mansfield
#' Park". The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/141}
#' @format A character vector with 14768 elements
"mansfieldpark"

#' The text of Jane Austen's novel "Emma"
#'
#' A dataset containing the text of Jane Austen's 1815 novel "Emma".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of about 70 characters each. 
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/158}
#' @format A character vector with 15297 elements
"emma"

#' The text of Jane Austen's novel "Northanger Abbey"
#'
#' A dataset containing the text of Jane Austen's novel "Northanger Abbey", 
#' published posthumously in 1818.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/121}
#' @format A character vector with 7840 elements
"northangerabbey"

#' The text of Jane Austen's novel "Persuasion"
#'
#' A dataset containing the text of Jane Austen's novel "Persuasion", 
#' published posthumously in 1818.
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/105}
#' @format A character vector with 8328 elements
"persuasion"


R/wilde_plays.R
#' Tidy data frame of Oscar Wilde's 7 completed, presented plays
#' 
#' Returns a tidy data frame of Oscar Wilde's 7 completed, presented plays with 
#' two columns: \code{text}, which contains the text of the plays divided into 
#' elements of up to about 70 characters each, and \code{play}, which contains
#' the titles of the plays as a factor in order of publication.
#' 
#' @details Users should be aware that there are some differences in usage 
#' between the novels as made available by Project Gutenberg. For example, 
#' "anything" vs. "any thing", "Mr" vs. "Mr.", and using underscores vs. all 
#' caps to indicate italics/emphasis.
#' 
#' @return A data frame with two columns: \code{text} and \code{play}
#' 
#' @name wilde_plays
#' 
#' @examples 
#' 
#' library(dplyr)
#'
#' wilde_plays() %>% 
#'     group_by(play) %>%
#'     summarise(total_lines = n())
#'
#' @export
wilde_plays <- function(){
        plays <- list(
                "Vera; Or, The Nihilists" = oscarwilder::vera_or_the_nihilists,
                "Salome" = oscarwilder::salome,
                "The Dutchess of Padua" = oscarwilder::the_dutchess_of_padua,
                "Lady Windermere's Fan" = oscarwilder::lady_windermeres_fan,
                "A Woman of No Importance" = oscarwilder::a_woman_of_no_importance,
                "An Ideal Husband" = oscarwilder::an_ideal_husband,
                "The Importance of Being Earnest" = oscarwilder::the_importance_of_being_earnest
        )
       ret <- data.frame(text = unlist(plays, use.names = FALSE), 
                         stringsAsFactors = FALSE)
       ret$play <- factor(rep(names(plays), sapply(plays, length)))
       ret$play <- factor(ret$play, levels = unique(ret$play))
       structure(ret, class = c("tbl_df", "tbl", "data.frame"))
}


R/austen_books.R
#' Tidy data frame of Jane Austen's 6 completed, published novels
#' 
#' Returns a tidy data frame of Jane Austen's 6 completed, published novels with 
#' two columns: \code{text}, which contains the text of the novels divided into 
#' elements of up to about 70 characters each, and \code{book}, which contains the titles of
#' the novels as a factor in order of publication.
#' 
#' @details Users should be aware that there are some differences in usage 
#' between the novels as made available by Project Gutenberg. For example, 
#' "anything" vs. "any thing", "Mr" vs. "Mr.", and using underscores vs. all 
#' caps to indicate italics/emphasis.
#' 
#' @return A data frame with two columns: \code{text} and \code{book}
#' 
#' @name austen_books
#' 
#' @examples 
#' 
#' library(dplyr)
#'
#' austen_books() %>% group_by(book) %>%
#'      summarise(total_lines = n())
#'
#' @export
austen_books <- function(){
        books <- list(
                "Sense & Sensibility" = sensesensibility,
                "Pride & Prejudice" = prideprejudice,
                "Mansfield Park" = mansfieldpark,
                "Emma" = emma,
                "Northanger Abbey" = northangerabbey,
                "Persuasion" = persuasion
        )
       ret <- data.frame(text = unlist(books, use.names = FALSE), 
                         stringsAsFactors = FALSE)
       ret$book <- factor(rep(names(books), sapply(books, length)))
       ret$book <- factor(ret$book, levels = unique(ret$book))
       structure(ret, class = c("tbl_df", "tbl", "data.frame"))
}


data-raw/prep_data.R
## Obtain texts from Gutenberg Project
library(gutenbergr)
vera_or_the_nihilists <- gutenberg_download(26494)$text
salome <- gutenberg_download(42704)$text
the_dutchess_of_padua <- gutenberg_download(875)$text
lady_windermeres_fan <- gutenberg_download(790)$text
a_woman_of_no_importance<- gutenberg_download(854)$text
an_ideal_husband <- gutenberg_download(885)$text
the_importance_of_being_earnest <- gutenberg_download(844)$text

library(stringr)
vera_or_the_nihilists <- str_trim(vera_or_the_nihilists, side = "both")
salome <- str_trim(salome, side = "both")
the_dutchess_of_padua <- str_trim(the_dutchess_of_padua, side = "both")
lady_windermeres_fan <- str_trim(lady_windermeres_fan, side = "both")
a_woman_of_no_importance <- str_trim(a_woman_of_no_importance, side = "both")
an_ideal_husband <- str_trim(an_ideal_husband, side = "both")
the_importance_of_being_earnest <- str_trim(the_importance_of_being_earnest, side = "both")

# Remove illustrations and cover references
salome <- salome[6:length(salome)]

## Now, add the data files to the package
usethis::use_data(vera_or_the_nihilists, overwrite = TRUE)
usethis::use_data(salome, overwrite = TRUE)
usethis::use_data(lady_windermeres_fan, overwrite = TRUE)
usethis::use_data(a_woman_of_no_importance, overwrite = TRUE)
usethis::use_data(the_importance_of_being_earnest, overwrite = TRUE)
usethis::use_data(an_ideal_husband, overwrite = TRUE)
usethis::use_data(the_dutchess_of_padua, overwrite = TRUE)


R/plays.R
#' Oscar Wilde's Complete Plays
#' 
#' This package contains the complete text of Oscar Wilde's 7 completed, 
#' published plays, formatted to be convenient for text analysis.
#' @docType package
#' @name oscarwilder
#' @aliases oscarwilder oscarwilder-package
NULL

#' The text of Oscar Wilde's play "Vera; Or, The Nihilists" (1880)
#'
#' A dataset containing the text of Oscar Wilde's 1880 play "Vera; Or, The
#' Nihilists".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/26494}
#' @format A character vector with 2835 elements
"vera_or_the_nihilists"

#' The text of Oscar Wilde's play "Salome" (1891)
#'
#' A dataset containing the text of Oscar Wilde's 1891 play "Salome".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/42704}
#' @format A character vector with 2522 elements
"salome"

#' The text of Oscar Wilde's play "The Dutchess of Padua" (1891)
#'
#' A dataset containing the text of Oscar Wilde's 1891 play "The Dutchess of
#' Padua".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/875}
#' @format A character vector with 5470 elements
"the_dutchess_of_padua"

#' The text of Oscar Wilde's play "Lady Windermere's Fan" (1892)
#'
#' A dataset containing the text of Oscar Wilde's 1892 play "Lady Windermere's 
#' Fan".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/790}
#' @format A character vector with 2912 elements
"lady_windermeres_fan"

#' The text of Oscar Wilde's play "A Woman of No Importance" (1893)
#'
#' A dataset containing the text of Oscar Wilde's 1893 play A Woman of "No 
#' Importance".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/854}
#' @format A character vector with 3274 elements
"a_woman_of_no_importance"

#' The text of Oscar Wilde's play "An Ideal Husband" (1895)
#'
#' A dataset containing the text of Oscar Wilde's 1893 play ""An Ideal Husband".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/885}
#' @format A character vector with 4464 elements
"an_ideal_husband"

#' The text of Oscar Wilde's play "The Importance of Being Earnest" (1895)
#'
#' A dataset containing the text of Oscar Wilde's 1893 play "The Importance of
#' Being Earnest".
#' The UTF-8 plain text was sourced from Project Gutenberg
#' and is divided into elements of up to about 70 characters each.
#' (Some elements are blank.)
#'
#' @source \url{http://www.gutenberg.org/ebooks/854}
#' @format A character vector with 3091 elements
"the_importance_of_being_earnest"
